File: implementations/cpp/edit_distance_and_alignment.cpp
/**
 * Reference Implementation: Edit Distance and Sequence Alignment
 * Demonstrates:
 * 1. Levenshtein Distance: Classical 2D DP (O(nm) Time, O(nm) Space) & Two-Row Rolling Array (O(m) Space).
 * 2. Full Edit Transcript Backtracing (Match, Substitute, Insert, Delete).
 * 3. Needleman-Wunsch Global Alignment: 2D DP Scoring & Two-Row Rolling Array.
 * 4. Needleman-Wunsch Alignment Backtracing (Optimal Aligned Strings with Gap Characters '-').
 * 5. Hamming Distance for Equal-Length Strings (O(n) Time, O(1) Space).
 *
 * Language: C++17
 */

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cassert>
#include <stdexcept>

namespace sequence_alignment {

// ============================================================================
// 1. Levenshtein Edit Distance (Cost Minimization)
// ============================================================================

/**
 * Classical 2D DP: O(nm) Time, O(nm) Space.
 * dp[i][j] = min edits to transform a[0..i-1] into b[0..j-1].
 */
int levenshtein_distance(const std::string& a, const std::string& b) {
    int n = static_cast<int>(a.size());
    int m = static_cast<int>(b.size());

    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(m + 1, 0));

    for (int i = 0; i <= n; ++i) dp[i][0] = i;
    for (int j = 0; j <= m; ++j) dp[0][j] = j;

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            int cost = (a[i - 1] == b[j - 1]) ? 0 : 1;
            dp[i][j] = std::min({
                dp[i - 1][j] + 1,      // delete from a
                dp[i][j - 1] + 1,      // insert into a
                dp[i - 1][j - 1] + cost // match or substitute
            });
        }
    }

    return dp[n][m];
}

/**
 * Two-Row Rolling Array Space Optimization: O(nm) Time, O(m) Space.
 */
int levenshtein_distance_rolling(const std::string& a, const std::string& b) {
    int n = static_cast<int>(a.size());
    int m = static_cast<int>(b.size());

    std::vector<int> prev(m + 1), cur(m + 1);
    for (int j = 0; j <= m; ++j) prev[j] = j;

    for (int i = 1; i <= n; ++i) {
        cur[0] = i;
        for (int j = 1; j <= m; ++j) {
            int cost = (a[i - 1] == b[j - 1]) ? 0 : 1;
            cur[j] = std::min({
                prev[j] + 1,
                cur[j - 1] + 1,
                prev[j - 1] + cost
            });
        }
        std::swap(prev, cur);
    }

    return prev[m];
}

// ============================================================================
// 2. Edit Transcript Reconstruction
// ============================================================================

struct EditOp {
    std::string type; // "match", "substitute", "insert", "delete"
    char from_char;
    char to_char;
};

std::pair<int, std::vector<EditOp>> levenshtein_with_ops(
    const std::string& a,
    const std::string& b) {

    int n = static_cast<int>(a.size());
    int m = static_cast<int>(b.size());

    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(m + 1, 0));

    for (int i = 0; i <= n; ++i) dp[i][0] = i;
    for (int j = 0; j <= m; ++j) dp[0][j] = j;

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            int cost = (a[i - 1] == b[j - 1]) ? 0 : 1;
            dp[i][j] = std::min({
                dp[i - 1][j] + 1,
                dp[i][j - 1] + 1,
                dp[i - 1][j - 1] + cost
            });
        }
    }

    std::vector<EditOp> ops;
    int i = n, j = m;

    while (i > 0 || j > 0) {
        if (i > 0 && j > 0) {
            int cost = (a[i - 1] == b[j - 1]) ? 0 : 1;
            if (dp[i][j] == dp[i - 1][j - 1] + cost) {
                if (cost == 0) {
                    ops.push_back({"match", a[i - 1], b[j - 1]});
                } else {
                    ops.push_back({"substitute", a[i - 1], b[j - 1]});
                }
                --i;
                --j;
                continue;
            }
        }

        if (i > 0 && dp[i][j] == dp[i - 1][j] + 1) {
            ops.push_back({"delete", a[i - 1], '\0'});
            --i;
        } else {
            ops.push_back({"insert", '\0', b[j - 1]});
            --j;
        }
    }

    std::reverse(ops.begin(), ops.end());
    return {dp[n][m], ops};
}

// ============================================================================
// 3. Needleman-Wunsch Global Alignment (Score Maximization)
// ============================================================================

/**
 * Computes optimal global alignment score: O(nm) Time, O(nm) Space.
 */
int needleman_wunsch_score(
    const std::string& a,
    const std::string& b,
    int match_score = 1,
    int mismatch_penalty = -1,
    int gap_penalty = -1) {

    int n = static_cast<int>(a.size());
    int m = static_cast<int>(b.size());

    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(m + 1, 0));

    for (int i = 0; i <= n; ++i) dp[i][0] = i * gap_penalty;
    for (int j = 0; j <= m; ++j) dp[0][j] = j * gap_penalty;

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            int diag = dp[i - 1][j - 1] + (a[i - 1] == b[j - 1] ? match_score : mismatch_penalty);
            int up = dp[i - 1][j] + gap_penalty;
            int left = dp[i][j - 1] + gap_penalty;
            dp[i][j] = std::max({diag, up, left});
        }
    }

    return dp[n][m];
}

/**
 * Two-Row Rolling Array for Needleman-Wunsch: O(nm) Time, O(m) Space.
 */
int needleman_wunsch_score_rolling(
    const std::string& a,
    const std::string& b,
    int match_score = 1,
    int mismatch_penalty = -1,
    int gap_penalty = -1) {

    int n = static_cast<int>(a.size());
    int m = static_cast<int>(b.size());

    std::vector<int> prev(m + 1), cur(m + 1);

    for (int j = 0; j <= m; ++j) prev[j] = j * gap_penalty;

    for (int i = 1; i <= n; ++i) {
        cur[0] = i * gap_penalty;
        for (int j = 1; j <= m; ++j) {
            int diag = prev[j - 1] + (a[i - 1] == b[j - 1] ? match_score : mismatch_penalty);
            int up = prev[j] + gap_penalty;
            int left = cur[j - 1] + gap_penalty;
            cur[j] = std::max({diag, up, left});
        }
        std::swap(prev, cur);
    }

    return prev[m];
}

struct AlignmentResult {
    int score;
    std::string aligned_a;
    std::string aligned_b;
};

/**
 * Global Alignment Reconstruction with Gap Characters '-': O(nm) Time, O(nm) Space.
 */
AlignmentResult needleman_wunsch_align(
    const std::string& a,
    const std::string& b,
    int match_score = 1,
    int mismatch_penalty = -1,
    int gap_penalty = -1) {

    int n = static_cast<int>(a.size());
    int m = static_cast<int>(b.size());

    std::vector<std::vector<int>> dp(n + 1, std::vector<int>(m + 1, 0));

    for (int i = 0; i <= n; ++i) dp[i][0] = i * gap_penalty;
    for (int j = 0; j <= m; ++j) dp[0][j] = j * gap_penalty;

    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j <= m; ++j) {
            int diag = dp[i - 1][j - 1] + (a[i - 1] == b[j - 1] ? match_score : mismatch_penalty);
            int up = dp[i - 1][j] + gap_penalty;
            int left = dp[i][j - 1] + gap_penalty;
            dp[i][j] = std::max({diag, up, left});
        }
    }

    std::string aligned_a, aligned_b;
    int i = n, j = m;

    while (i > 0 || j > 0) {
        if (i > 0 && j > 0) {
            int diag = dp[i - 1][j - 1] + (a[i - 1] == b[j - 1] ? match_score : mismatch_penalty);
            if (dp[i][j] == diag) {
                aligned_a.push_back(a[i - 1]);
                aligned_b.push_back(b[j - 1]);
                --i;
                --j;
                continue;
            }
        }

        if (i > 0 && dp[i][j] == dp[i - 1][j] + gap_penalty) {
            aligned_a.push_back(a[i - 1]);
            aligned_b.push_back('-');
            --i;
        } else {
            aligned_a.push_back('-');
            aligned_b.push_back(b[j - 1]);
            --j;
        }
    }

    std::reverse(aligned_a.begin(), aligned_a.end());
    std::reverse(aligned_b.begin(), aligned_b.end());

    return {dp[n][m], aligned_a, aligned_b};
}

// ============================================================================
// 4. Comparative Metrics: Hamming Distance
// ============================================================================

/**
 * Counts point mismatches between equal-length sequences: O(n) Time, O(1) Space.
 */
int hamming_distance(const std::string& a, const std::string& b) {
    if (a.size() != b.size()) {
        throw std::invalid_argument("Hamming distance requires equal length strings");
    }
    int diff = 0;
    for (size_t i = 0; i < a.size(); ++i) {
        if (a[i] != b[i]) {
            ++diff;
        }
    }
    return diff;
}

} // namespace sequence_alignment

// ============================================================================
// Comprehensive Unit Verification
// ============================================================================

int main() {
    using namespace sequence_alignment;

    // Test 1: Canonical kitten -> sitting
    std::string s1 = "kitten", s2 = "sitting";
    assert(levenshtein_distance(s1, s2) == 3);
    assert(levenshtein_distance_rolling(s1, s2) == 3);

    auto [dist1, ops1] = levenshtein_with_ops(s1, s2);
    assert(dist1 == 3);
    // Verify ops transform s1 into s2
    int edit_count = 0;
    for (const auto& op : ops1) {
        if (op.type != "match") ++edit_count;
    }
    assert(edit_count == 3);

    // Test 2: horse -> ros (Arthur's worked example)
    std::string s3 = "horse", s4 = "ros";
    assert(levenshtein_distance(s3, s4) == 3);
    assert(levenshtein_distance_rolling(s3, s4) == 3);

    // Test 3: Empty string edge cases
    assert(levenshtein_distance("", "") == 0);
    assert(levenshtein_distance("abc", "") == 3);
    assert(levenshtein_distance("", "abcd") == 4);
    assert(levenshtein_distance_rolling("", "abcd") == 4);

    // Test 4: Identical strings
    assert(levenshtein_distance("algorithm", "algorithm") == 0);
    assert(levenshtein_distance_rolling("algorithm", "algorithm") == 0);

    // Test 5: Needleman-Wunsch Global Alignment
    // a = GATTACA, b = GCATGCU
    std::string dna1 = "GATTACA", dna2 = "GCATGCU";
    int score = needleman_wunsch_score(dna1, dna2, 1, -1, -1);
    int rolling_score = needleman_wunsch_score_rolling(dna1, dna2, 1, -1, -1);
    assert(score == rolling_score);

    auto align_res = needleman_wunsch_align(dna1, dna2, 1, -1, -1);
    assert(align_res.score == score);
    assert(align_res.aligned_a.size() == align_res.aligned_b.size());

    // Verify score from aligned strings manually
    int manual_score = 0;
    for (size_t k = 0; k < align_res.aligned_a.size(); ++k) {
        char ca = align_res.aligned_a[k];
        char cb = align_res.aligned_b[k];
        if (ca == '-' || cb == '-') {
            manual_score += -1; // gap
        } else if (ca == cb) {
            manual_score += 1;  // match
        } else {
            manual_score += -1; // mismatch
        }
    }
    assert(manual_score == score);

    // Test 6: Hamming Distance
    assert(hamming_distance("karolin", "kathrin") == 3);
    assert(hamming_distance("1011101", "1001001") == 2);
    try {
        hamming_distance("short", "longer");
        assert(false);
    } catch (const std::invalid_argument&) {}

    std::cout << "[PASS] All Edit Distance and Sequence Alignment C++ unit tests passed." << std::endl;
    return 0;
}

File: implementations/python/edit_distance_and_alignment.py
"""
Reference Implementation: Edit Distance and Sequence Alignment
Demonstrates:
1. Levenshtein Distance: Classical 2D DP (O(nm) Time, O(nm) Space) & Two-Row Rolling Array (O(m) Space).
2. Full Edit Transcript Backtracing (Match, Substitute, Insert, Delete).
3. Needleman-Wunsch Global Alignment: 2D DP Scoring & Two-Row Rolling Array.
4. Needleman-Wunsch Alignment Backtracing (Optimal Aligned Strings with Gap Characters '-').
5. Hamming Distance for Equal-Length Strings (O(n) Time, O(1) Space).

Language: Python 3
"""

import unittest
from typing import List, Tuple, Optional


# ============================================================================
# 1. Levenshtein Edit Distance (Cost Minimization)
# ============================================================================

def levenshtein_distance(a: str, b: str) -> int:
    """
    Classical 2D DP: O(nm) Time, O(nm) Space.
    dp[i][j] = min edits to transform a[0..i-1] into b[0..j-1].
    """
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    for i in range(n + 1):
        dp[i][0] = i
    for j in range(m + 1):
        dp[0][j] = j

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            dp[i][j] = min(
                dp[i - 1][j] + 1,      # delete from a
                dp[i][j - 1] + 1,      # insert into a
                dp[i - 1][j - 1] + cost # match or substitute
            )

    return dp[n][m]


def levenshtein_distance_rolling(a: str, b: str) -> int:
    """
    Two-Row Rolling Array Space Optimization: O(nm) Time, O(m) Space.
    """
    n, m = len(a), len(b)
    prev = list(range(m + 1))
    cur = [0] * (m + 1)

    for i in range(1, n + 1):
        cur[0] = i
        for j in range(1, m + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            cur[j] = min(
                prev[j] + 1,
                cur[j - 1] + 1,
                prev[j - 1] + cost
            )
        prev, cur = cur, prev

    return prev[m]


# ============================================================================
# 2. Edit Transcript Reconstruction
# ============================================================================

def levenshtein_with_ops(a: str, b: str) -> Tuple[int, List[Tuple[str, Optional[str], Optional[str]]]]:
    """
    Reconstructs the full sequence of edit operations from a to b.
    Returns (distance, ops) where each op is (type, from_char, to_char).
    """
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    for i in range(n + 1):
        dp[i][0] = i
    for j in range(m + 1):
        dp[0][j] = j

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            dp[i][j] = min(
                dp[i - 1][j] + 1,
                dp[i][j - 1] + 1,
                dp[i - 1][j - 1] + cost
            )

    ops = []
    i, j = n, m

    while i > 0 or j > 0:
        if i > 0 and j > 0:
            cost = 0 if a[i - 1] == b[j - 1] else 1
            if dp[i][j] == dp[i - 1][j - 1] + cost:
                if cost == 0:
                    ops.append(("match", a[i - 1], b[j - 1]))
                else:
                    ops.append(("substitute", a[i - 1], b[j - 1]))
                i -= 1
                j -= 1
                continue

        if i > 0 and dp[i][j] == dp[i - 1][j] + 1:
            ops.append(("delete", a[i - 1], None))
            i -= 1
        else:
            ops.append(("insert", None, b[j - 1]))
            j -= 1

    ops.reverse()
    return dp[n][m], ops


# ============================================================================
# 3. Needleman-Wunsch Global Alignment (Score Maximization)
# ============================================================================

def needleman_wunsch_score(
    a: str,
    b: str,
    match_score: int = 1,
    mismatch_penalty: int = -1,
    gap_penalty: int = -1
) -> int:
    """
    Computes optimal global alignment score: O(nm) Time, O(nm) Space.
    """
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    for i in range(n + 1):
        dp[i][0] = i * gap_penalty
    for j in range(m + 1):
        dp[0][j] = j * gap_penalty

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            diag = dp[i - 1][j - 1] + (match_score if a[i - 1] == b[j - 1] else mismatch_penalty)
            up = dp[i - 1][j] + gap_penalty
            left = dp[i][j - 1] + gap_penalty
            dp[i][j] = max(diag, up, left)

    return dp[n][m]


def needleman_wunsch_score_rolling(
    a: str,
    b: str,
    match_score: int = 1,
    mismatch_penalty: int = -1,
    gap_penalty: int = -1
) -> int:
    """
    Two-Row Rolling Array for Needleman-Wunsch: O(nm) Time, O(m) Space.
    """
    n, m = len(a), len(b)
    prev = [j * gap_penalty for j in range(m + 1)]
    cur = [0] * (m + 1)

    for i in range(1, n + 1):
        cur[0] = i * gap_penalty
        for j in range(1, m + 1):
            diag = prev[j - 1] + (match_score if a[i - 1] == b[j - 1] else mismatch_penalty)
            up = prev[j] + gap_penalty
            left = cur[j - 1] + gap_penalty
            cur[j] = max(diag, up, left)
        prev, cur = cur, prev

    return prev[m]


def needleman_wunsch_align(
    a: str,
    b: str,
    match_score: int = 1,
    mismatch_penalty: int = -1,
    gap_penalty: int = -1
) -> Tuple[int, str, str]:
    """
    Global Alignment Reconstruction with Gap Characters '-': O(nm) Time, O(nm) Space.
    Returns (score, aligned_a, aligned_b).
    """
    n, m = len(a), len(b)
    dp = [[0] * (m + 1) for _ in range(n + 1)]

    for i in range(n + 1):
        dp[i][0] = i * gap_penalty
    for j in range(m + 1):
        dp[0][j] = j * gap_penalty

    for i in range(1, n + 1):
        for j in range(1, m + 1):
            diag = dp[i - 1][j - 1] + (match_score if a[i - 1] == b[j - 1] else mismatch_penalty)
            up = dp[i - 1][j] + gap_penalty
            left = dp[i][j - 1] + gap_penalty
            dp[i][j] = max(diag, up, left)

    aligned_a = []
    aligned_b = []
    i, j = n, m

    while i > 0 or j > 0:
        if i > 0 and j > 0:
            diag = dp[i - 1][j - 1] + (match_score if a[i - 1] == b[j - 1] else mismatch_penalty)
            if dp[i][j] == diag:
                aligned_a.append(a[i - 1])
                aligned_b.append(b[j - 1])
                i -= 1
                j -= 1
                continue

        if i > 0 and dp[i][j] == dp[i - 1][j] + gap_penalty:
            aligned_a.append(a[i - 1])
            aligned_b.append('-')
            i -= 1
        else:
            aligned_a.append('-')
            aligned_b.append(b[j - 1])
            j -= 1

    aligned_a.reverse()
    aligned_b.reverse()
    return dp[n][m], ''.join(aligned_a), ''.join(aligned_b)


# ============================================================================
# 4. Comparative Metrics: Hamming Distance
# ============================================================================

def hamming_distance(a: str, b: str) -> int:
    """
    Counts point mismatches between equal-length sequences: O(n) Time, O(1) Space.
    """
    if len(a) != len(b):
        raise ValueError("Hamming distance requires equal length strings")
    return sum(ca != cb for ca, cb in zip(a, b))


# ============================================================================
# Unit Tests
# ============================================================================

class TestEditDistanceAndAlignment(unittest.TestCase):
    def test_levenshtein_kitten_sitting(self):
        s1, s2 = "kitten", "sitting"
        self.assertEqual(levenshtein_distance(s1, s2), 3)
        self.assertEqual(levenshtein_distance_rolling(s1, s2), 3)

        dist, ops = levenshtein_with_ops(s1, s2)
        self.assertEqual(dist, 3)
        non_matches = [op for op in ops if op[0] != "match"]
        self.assertEqual(len(non_matches), 3)

    def test_levenshtein_horse_ros(self):
        s1, s2 = "horse", "ros"
        self.assertEqual(levenshtein_distance(s1, s2), 3)
        self.assertEqual(levenshtein_distance_rolling(s1, s2), 3)

    def test_empty_and_identical(self):
        self.assertEqual(levenshtein_distance("", ""), 0)
        self.assertEqual(levenshtein_distance("abc", ""), 3)
        self.assertEqual(levenshtein_distance("", "abcd"), 4)
        self.assertEqual(levenshtein_distance("algorithm", "algorithm"), 0)
        self.assertEqual(levenshtein_distance_rolling("algorithm", "algorithm"), 0)

    def test_needleman_wunsch(self):
        dna1, dna2 = "GATTACA", "GCATGCU"
        score = needleman_wunsch_score(dna1, dna2, 1, -1, -1)
        rolling = needleman_wunsch_score_rolling(dna1, dna2, 1, -1, -1)
        self.assertEqual(score, rolling)

        align_score, a_str, b_str = needleman_wunsch_align(dna1, dna2, 1, -1, -1)
        self.assertEqual(align_score, score)
        self.assertEqual(len(a_str), len(b_str))

        # Manual validation
        calc = 0
        for ca, cb in zip(a_str, b_str):
            if ca == '-' or cb == '-':
                calc += -1
            elif ca == cb:
                calc += 1
            else:
                calc += -1
        self.assertEqual(calc, score)

    def test_hamming_distance(self):
        self.assertEqual(hamming_distance("karolin", "kathrin"), 3)
        self.assertEqual(hamming_distance("1011101", "1001001"), 2)
        with self.assertRaises(ValueError):
            hamming_distance("abc", "abcd")


if __name__ == '__main__':
    unittest.main()
